src/Docfx.Common/Git/GitUtility.cs

```diff
@@ -32,73 +24,97 @@ public static class GitUtility
 
-    /// <summary>
-    /// Gets git is installed globally or not.
-    /// </summary>
-    public static Lazy<bool> ExistGitCommand = new(() =>
-    {
-        try
-        {
-            bool gitCommandExists = CommandUtility.ExistCommand(CommandName);
-            if (!gitCommandExists)
-            {
-                Logger.LogInfo("Looks like Git is not installed globally. We depend on Git to extract repository information for source code and files.");
-            }
-            return gitCommandExists;
-        }
-        catch(Exception ex)
-        {
-            Logger.LogWarning("Failed to get Git command that installed globally. Exception: " + ex.ToString());
-            return false;
-        }
-    }, LazyThreadSafetyMode.ExecutionAndPublication);
-
-    public static GitDetail TryGetFileDetail(string filePath)
+    public static GitDetail? TryGetFileDetail(string filePath)
     {
         if (EnvironmentContext.GitFeaturesDisabled)
             return null;
 
-        try
-        {
-            return GetFileDetail(filePath);
-        }
-        catch (Exception ex)
-        {
-            Logger.LogWarning($"Skipping GetFileDetail. Exception found: {ex.GetType()}, Message: {ex.Message}");
-            Logger.LogVerbose(ex.ToString());
-        }
+        var repo = GetRepoInfo(Path.GetDirectoryName(filePath));
+        if (repo is null)
+            return null;
 
-        return null;
+        return new()
+        {
+            Repo = repo.url,
+            Branch = repo.branch,
+            Path = Path.GetRelativePath(repo.path, filePath),
+        };
     }
 
-    public static string RawContentUrlToContentUrl(string rawUrl)
+    public static string? RawContentUrlToContentUrl(string rawUrl)
     {
         if (EnvironmentContext.GitFeaturesDisabled)
             return null;
 
-        var branch = Environment.GetEnvironmentVariable("DOCFX_SOURCE_BRANCH_NAME");
-
         // GitHub
         return Regex.Replace(
             rawUrl,
             @"^https://raw\.githubusercontent\.com/([^/]+)/([^/]+)/([^/]+)/(.+)$",
-            string.IsNullOrEmpty(branch) ? "https://github.com/$1/$2/blob/$3/$4" : $"https://github.com/$1/$2/blob/{branch}/$4");
+            string.IsNullOrEmpty(s_branch) ? "https://github.com/$1/$2/blob/$3/$4" : $"https://github.com/$1/$2/blob/{s_branch}/$4");
+    }
+
+    private static Repo? GetRepoInfo(string? directory)
+    {
+        if (string.IsNullOrEmpty(directory))
+            return null;
+
+        return s_cache.GetOrAdd(directory, _ =>
+        {
+            if (IsGitRoot(directory))
+            {
+                return GetRepoInfoCore(directory).Result;
+            }
+
+            return GetRepoInfo(Path.GetDirectoryName(directory));
+        });
+
+        static async Task<Repo?> GetRepoInfoCore(string directory)
+        {
+            using var repo = await Repository.Factory.OpenPrimitiveAsync(directory);
```

Comment:
It seems not works on `git submodule` path.
`Repository does not exist.` exception occurred at following lines.

https://github.com/kekyo/GitReader/blob/4126738704ee8b6e330e99c6bc81e9de8bc925c9/GitReader.Core/Primitive/PrimitiveRepositoryFacade.cs#L27-L30